File: 23.MergekSortedLists.py
import heapq
from typing import List, Tuple


class ListNode:
    def __str__(self) -> str:
        res = []
        res.append(str(self.val))
        while self.next:  # type:ignore
            res.append(str(self.next.val))  # type:ignore
            self.next = self.next.next  # type:ignore
        return "->".join(res)

    def __init__(self, x: int) -> None:
        self.val = x
        self.next = None

    # def __lt__(self, other):
    #     return self.val < other.val


class Solution:
    def mergeKLists(self, lists: List[ListNode]) -> ListNode:
        ListNode.__lt__ = lambda self, other: self.val < other.val  # type:ignore
        data: List[Tuple[int, ListNode]] = []
        dumpy = ListNode(0)
        begin = dumpy
        heapq.heapify(data)
        for l in lists:
            if l:
                heapq.heappush(data, (l.val, l))
        while data:
            val, head = heapq.heappop(data)
            dumpy.next = head
            dumpy = dumpy.next
            if head.next:
                heapq.heappush(data, (head.next.val, head.next))

        return begin.next


if __name__ == "__main__":
    #   1->4->5,
    # 1->3->4,
    # 2->6
    l11 = ListNode(1)
    l12 = ListNode(4)
    l13 = ListNode(5)
    l21 = ListNode(1)
    l22 = ListNode(3)
    l23 = ListNode(4)
    l31 = ListNode(2)
    l32 = ListNode(6)
    l11.next = l12  # type: ignore
    l12.next = l13  # type: ignore
    l21.next = l22  # type: ignore
    l22.next = l23  # type: ignore
    l31.next = l32  # type: ignore
    s = Solution()
    print(s.mergeKLists([l11, l21, l31]))
    # 1->1->2->3->4->4->5->6

File: 23.MergekSortedLists.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type ListNode struct {
	Val  int
	Next *ListNode
}

func (l *ListNode) String() string {
	res := []string{}
	for l != nil {
		res = append(res, strconv.Itoa(l.Val))
		l = l.Next
	}
	return strings.Join(res, "->")
}

func mergeKLists(lists []*ListNode) *ListNode {
	return merge(lists, 0, len(lists)-1)

}

func merge(lists []*ListNode, l, r int) *ListNode {
	if l > r {
		return nil
	}
	if l == r {
		return lists[l]
	}
	if l+1 == r {
		return mergeTwoLists(lists[l], lists[r])
	}
	m := (l + r) / 2
	l1 := merge(lists, l, m)
	l2 := merge(lists, m+1, r)
	return mergeTwoLists(l1, l2)
}

func mergeTwoLists(l1, l2 *ListNode) *ListNode {
	tail := &ListNode{}
	dumpy := tail
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			tail.Next = l1
			l1 = l1.Next
		} else {
			tail.Next = l2
			l2 = l2.Next
		}
		tail = tail.Next
	}
	if l1 != nil {
		tail.Next = l1
	}
	if l2 != nil {
		tail.Next = l2
	}
	return dumpy.Next
}

func main() {
	// 1->4->5,
	// 1->3->4,
	// 2->6
	l11 := &ListNode{Val: 1}
	l12 := &ListNode{Val: 4}
	l13 := &ListNode{Val: 5}
	l11.Next = l12
	l12.Next = l13
	l21 := &ListNode{Val: 1}
	l22 := &ListNode{Val: 3}
	l23 := &ListNode{Val: 4}
	l21.Next = l22
	l22.Next = l23
	l31 := &ListNode{Val: 2}
	l32 := &ListNode{Val: 6}
	l31.Next = l32
	// 1->1->2->3->4->4->5->6
	fmt.Println("l1", l11)
	// fmt.Println("two lists", mergeTwoLists(l11, l21))
	fmt.Println(mergeKLists([]*ListNode{l11, l21, l31}))
}
